### DISCUSSION TITLE
Road to 1.0

### DISCUSSION THREAD

**Discussion by stephencelis** [pointfreeco/swift-composable-architecture#1477] [Beta]
It’s been almost 2 and a half years since we first released the Composable Architecture, so we’ve been getting more and more questions about its 0.x.x prerelease versioning, what’s holding up a 1.0, and (mostly jokingly) whether it’s “ready for production” yet 😆

We thought it’d be a good opportunity to lay out our roadmap, and detail our plans for rolling out some of the first breaking changes we’d like to introduce in preparation of the release.

## What’s missing?

We’ve held off on 1.0 for this long not because we believe the library is in beta/unstable territory, but because there were a few key problem areas of library use that came up again and again, and we wanted to provide the nicest built-in solutions to them before considering the Composable Library a “complete” package.

We’ve slowly chipped away at these problem areas, which led to two significant public betas this year:

- [Concurrency](https://github.com/pointfreeco/swift-composable-architecture/discussions/1186): We wanted to make working with effects in the Composable Architecture much more approachable. Rather than require library-level use of Combine, which is powerful but comes with a significant learning curve, we introduced a friendlier set of async-await APIs.

- [Protocol](https://github.com/pointfreeco/swift-composable-architecture/discussions/1282): We finally merged this feature on Monday, and it introduces a friendlier, more modern interface for building features in the Composable Architecture.

So...what's left!?

### Navigation

We think the final “missing” feature is a holistic story for navigation. Navigation is a significant part of any application, and while the Composable Architecture has shipped with demonstrations of navigation from its first release, we’ve been far from happy with them.

So we’ve been experimenting with better navigation tools from day one, and over time have made significant strides. SwiftUI 3 completely revamped its stack navigation, so we’ve done our best to digest these changes and intend to ship tools that fully embrace all of SwiftUI’s navigation idioms, and even improve upon them.

So Navigation is our next and final focus before shipping a 1.0. We still have some experimentation and polish to do, but when we’re ready we’ll have another beta period waiting for you!

### Other in-flight work

We have many ongoing experiments in the Composable Architecture that we do not consider blockers for 1.0. Some of these features will make their way in before 1.0 is released. This includes work on a "non-exhaustive" version of the test store [first demonstrated](https://www.merowing.info/exhaustive-testing-in-tca/) by @krzysztofzablocki.

## Breaking changes

Throughout the 0.x.x period, we’ve done our best to maintain source compatibility across releases, and we’ve managed a [large catalog of deprecations](https://github.com/pointfreeco/swift-composable-architecture/blob/e4b3cfaf1fa243853a15cbe1fda16843ca10a037/Sources/ComposableArchitecture/Internal/Deprecations.swift) over many many releases. This means that all changes have been additive, even changes that have been complete reimaginings of existing features, like the `ReducerProtocol`. 1.0 gives us the opportunity to shed off some of these deprecated features, and tidy up the new features in the process.

There are two major changes we’d like to make that will require a few releases and some churn for our users, but we hope that you all find it worth it!

### Removing the old `Reducer` struct

The `ReducerProtocol` is now the de facto way of building reducers in the Composable Architecture. Alongside its release, we renamed the existing `Reducer` struct to `AnyReducer`, but provided a soft-deprecated type alias for compatibility.

`ReducerProtocol` is a mouthful, though, so we’d like to take steps over a few releases to fully deprecate the old `Reducer` struct and finally rename `ReducerProtocol` to `Reducer`. We will accomplish this in the following releases: 

1. Very soon we plan on shipping 0.43.0 and hard-deprecating the [`Reducer` type alias](https://github.com/pointfreeco/swift-composable-architecture/blob/e4b3cfaf1fa243853a15cbe1fda16843ca10a037/Sources/ComposableArchitecture/Reducer/AnyReducer/AnyReducer.swift#L41-L52). This will introduce a number of warnings to projects that have not yet migrated to the `ReducerProtocol`. To fix these warnings, you can simply rename all instances of `Reducer` to `AnyReducer`, and Xcode should be able to help a lot with that.

2. After 0.43.0, we will introduce a release candidate branch for 1.0 that will directly track the most recent release in the library with some additional, bigger breaking changes. Targeting this branch in your project will allow you to get a head start on being fully 1.0-compatible.

    The changes include:

    1. Removing the deprecated `typealias Reducer = AnyReducer`
    2. Renaming `ReducerProtocol` to `Reducer`
    3. Introducing a hard-deprecated `typealias ReducerProtocol = Reducer`

    If you have already updated your code base to 0.43.0, then none of these changes are breaking. You will just have to rename occurrences of `ReducerProtocol` to `Reducer`, which Xcode should be able to help with.

3. Closer to 1.0’s release, we will take steps to fully deprecate `AnyReducer` and its APIs.

    Our aim is to minimize friction, despite the above changes. So with 1.0 we do not plan on leaving folks that rely on `AnyReducer` behind. Instead, we plan on allowing folks to continue to use `AnyReducer`, but we will hide this deprecated type and its APIs from view, either as a companion module that you will need to import, or by utilizing `@_spi` imports to hide them from view and use unless specifically imported via `@_spi(AnyReducer) import ComposableArchitecture`.

### Removing many Combine-specific features of `Effect`

With the Concurrency release, we vastly simplified the `Effect` type, with a focus on 3 entry points for creating effects using Swift Concurrency (`.run`, `.task`, and `.fireAndForget`). In the process, we soft-deprecated much of the type’s existing Combine integration, and before 1.0, we’d like to remove as much of this legacy integration as possible. Folks that like using Combine can continue to do so by iterating over Combine’s `AsyncPublisher` async sequence in an `Effect.run`.

`Effect` was initially designed to be used as a Combine publisher, so it picked up some quirks from Combine that feel no longer necessary. For example, it picked up a `Failure` generic from the `Publisher` protocol, but this `Failure` is always locked to `Never` when used from within a reducer.

With the new entry points designed for Swift Concurrency, the `Failure` generic no longer makes sense, so we’d like to simplify `Effect<Action, Failure>` to `Effect<Action>`. We intend to do so over a number of releases, in a fashion similar to our planned rename of the `ReducerProtocol`:

1. With 0.43.0, we will rename `Effect` to `EffectPublisher` and provide a soft-deprecated typealias for `Effect = EffectPublisher` so that existing code still compiles. We will also introduce a typealias `EffectTask<Action> = Effect<Action, Never>`, which will be the new type you should start using in your reducers. We will also update the reducer signature to use `EffectTask` for simplicity.

2. Shortly after (in 0.44.0?), we will hard-deprecate the `Effect` type alias, which means those using `Effect<Action, Never>` can just rename to `EffectTask<Action>`, and those that use `Effect` with a non-`Never` failure can either take the opportunity to convert that code to async/await, or they can rename to `EffectPublisher`. 

3. And then, we will make the following changes to the prerelease 1.0 branch so that you can get a head start on updating your codebase to be 1.0-compliant:

    1. Remove the deprecated `Effect` type alias
    2. Rename `EffectTask` to `Effect`
    3. Introduce a deprecated `typealias EffectTask = Effect` so that all existing uses of `EffectTask` continue to work, and Xcode should be able to help make these changes for you.

4. Closer to 1.0’s release, we will take steps to fully deprecate `EffectPublisher` and its APIs. And similar to how we handled the removal of `AnyReducer` by using `@_spi`, we will hide `EffectPublisher` behind `@_spi` so that it is still accessible, but just with extra work of needing to specify `@_spi` when importing the Composable Architecture.
We want your input!

---

These changes are not insignificant, but we believe they pave the way to a simpler 1.0 that will be easier to introduce to your team mates and colleagues 🙂 

It’s worth mentioning that the beta period for our navigation tools will probably be comparable to what it was for the reducer protocol, and so that means the 1.0 prerelease branch will be around for quite awhile so that people should have plenty of time to update their codebases.

Please let us know if you see any issues with our strategies, or if you have any suggestions.

**Comment by mbrandonw** [comment]
We have another change we'd like to make, but it requires more input from the community.

We currently support iOS back to 13 as well as Swift 5.5+5.6. That comes with a decent amount of code complexity (we have lots of `#if swift` checks in the code), and we can't adopt things like `@StateObject` to optimize certain TCA constructs such as `WithViewStore`.

So, we'd like to drop support for iOS 13 and Swift 5.6 in the 1.0 release of TCA. Two things to keep in mind if you need to support iOS 13 or Swift 5.6:

* You will still have access to the new navigation tools as that will be in a 0.x.x release just before 1.0. You will just need to continue targeting the 0.x.x release until you can fully drop iOS 13/Swift 5.6.
* The 1.0 release isn't coming for another couple of months (probably in February or March of 2023).

So, is this going to be problematic for a large number of people?

**Reply by stephencelis** [reply]
Just a quick note, that while :-1: reactions provide sentiment, comments would be more helpful. If this change is problematic for you, please take the time to describe your requirements here. Is there anything keeping you on Swift 5.6? Will you still need iOS 13 support in 2023, and what percentage of your users are currently on iOS 13?

**Reply by msmollin-community** [reply]
Removing Swift 5.6 would drop support for Xcode 13, correct? I could see that being a larger issue if Apple hasn't sorted the issues with Xcode 14 by the time the 1.0 release is ready?

**Reply by mbrandonw** [reply]
Yeah, dropping Swift 5.6 means dropping Xcode 13 and lower.

Is that the bigger concern for people, or is it iOS 13?

If we need to split this step into two, first removing iOS 13 and then someday later removing Swift 5.6 we would be open to that.

**Reply by stephencelis** [reply]
@msmollin-community Given our target of Feburary/March 2023, and based off [past years of Xcode releases](https://xcodereleases.com), we're looking at having Xcode 14.3 by that point, and maybe even Swift 5.8. Hard to say what issues will be fixed/introduced in that time, but are there any particular ones that are keeping you on Xcode 13 today?

Also please note that the plan is that you will continue to be able to use 0.x.x releases, which will contain all the same features as 1.0.0.

**Reply by jshier** [reply]
And if they follow previous patterns, Apple will require Xcode 14 / iOS 16 SDK for App Store submissions in the spring, likely April. 

I'd say it'd be much easier to drop Swift 5.6 / Xcode 13 than iOS 13 for most people (who haven't dropped iOS 13 already).

**Reply by Jeehut** [reply]
No problem for me, and most teams I know are on current stable iOS + 1 or + 2 versions back. By the time you release this, this will mean that most teams will no longer need to support iOS 13 and for those who do, you have the 0.x release until they can make the switch, too. I think this breaking change in very sensible.

**Reply by larryonoff** [reply]
I assume that dropping iOS 13 (somewhere in February or March of 2023) isn't a good option. As for me the best option of dropping iOS 12 is in ~1-2 years. Somewhere after September release of new OS-es from Apple.

We dropped iOS 12 for almost all our apps recently, but we still have one app with iOS 12 support. We cannot drop iOS 12 since high percentage of paid users. All our apps use company's SDK, so we have to maintain our copy of TCA (BTW based on ReactiveSwift for the moment) with tons of @available. And we're planning to migrate to official TCA very soon.

Companies with big user-base cannot move so quick, so it would be great having support for a bit longer period than stated. As I mentioned after new all OS-es (iOS + iPadOS + macOS) releases.

**Reply by andriyslyusar** [reply]
In our company we have a requirement on maintaining 2 latest iOS major releases, so dropping iOS 13 and Xcode 13 are okey. I agree with @jshier that most likely by spring 2023 Apple will require use Xcode 14 for uploading apps to App Store.

**Reply by larryonoff** [reply]
Xcode 14 is not a big deal. Since it supports iOS 13.

**Reply by andreyz** [reply]
OK for both deprecations.

**Reply by mbrandonw** [reply]
> All our apps use company's SDK, so we have to maintain our copy of TCA (BTW based on ReactiveSwift for the moment) with tons of @available. And we're planning to migrate to official TCA very soon.

@larryonoff Considering that y'all have been maintaining a fork of TCA already to support iOS 12, it seems natural that when you can drop iOS 12 and convert to TCA proper you can maintain a fork of the last 0.x.x release that replays any significant changes you need from later 1.x.x releases. In fact, the fork could probably be exactly the 1.x.x release with just `@ObservedObject` swapped out for `@StateObject`, which would be a lot easier to do than what you have to go through now. It would even be a great community project for us to link to in the readme. Stephen and I could discuss maintaining it ourselves, but honestly it would probably be best in the hands of someone who actually needs that functionality.

And just to give some background details on this `@StateObject` thing. We want to support it for `WithViewStore` because it can greatly reduce the number of scoped stores and view stores created, which can be a performance improvement. We unfortunately cannot shim our own `@StateObject` because it seems SwiftUI does some magic under the hood that we haven't been able to recreate, _and_ we can't do availability checks because that leads to [`AnyView`s][view-builder-availability] which then comes with its own performance problems.

By holding onto iOS 13 longer we are shipping a less performant library to most people.

> Companies with big user-base cannot move so quick, so it would be great having support for a bit longer period than stated. As I mentioned after new all OS-es (iOS + iPadOS + macOS) releases.

If we drop support for iOS 13 in March in would mean we have supported 3.5 years of iOS versions, which seems reasonable. And to iterate for those who may have missed it, the 1.0 release will not introduce any new features whatsoever. It is exists only to make our first official breaking change in the library. The last 0.x.x release will have all of the navigation tools, and we currently are not even aware of any big features to come after 1.0.

[view-builder-availability]: https://developer.apple.com/documentation/swiftui/viewbuilder/buildlimitedavailability(_:)

**Reply by larryonoff** [reply]
Thanks for the details! Sounds reasonable.

**Reply by mbrandonw** [reply]
After discussing with @stephencelis we think we can meet halfway when it comes to dropping iOS 13.

First, we will _not_ drop iOS 13 support, but we will make `WithViewStore` available only on iOS 14 so that we can use `@StateObject`. That will be a breaking change for those using iOS 13, and to help those people we are considering one of the following:

* Recommend that people on iOS 13 use `@ObservedObject var viewStore` directly instead of using `WithViewStore`.
* Ship a `LegacyWithViewStore` that is basically the same as `WithViewStore`, but uses `@ObservedObject` under the hood.
* Make it possible for others to implement their own iOS 13 compatible `WithViewStore`. That would require us to make one single initializer public, which doesn't seem to be a problem.

So, if it turns out that truly the only reason we are dropping iOS 13 is to use `@StateObject`, then we think the above options will allow us to keep supporting iOS 13 without holding back everyone supporting 14+.

**Reply by larryonoff** [reply]
This really wise. BTW the first idea was to bump WithViewStore to iOS 14, like suggested already.

LegacyWithViewStore sounds good for me.

Would it be possible to use "if available" inside WithViewStore instead of bumping minimum OS versions?

**Reply by mbrandonw** [reply]
> Would it be possible to use "if available" inside WithViewStore instead of bumping minimum OS versions?

Unfortunately no. When you do `if #available` in a view builder it calls down to [`buildLimitedAvailability`](https://developer.apple.com/documentation/swiftui/viewbuilder/buildlimitedavailability(_:)), which returns an `AnyView`:

```swift
static func buildLimitedAvailability<Content>(_ content: Content) -> AnyView
```

And it makes sense that it needs to return `AnyView`, after all the iOS you are targeting is clearly doesn't have access to the static symbol you want to use. But, that `AnyView` actually comes with a measurable performance penalty which completely negates using the `@StateObject` in the first place.

**Reply by tgrapperon** [reply]
`WithViewStore` is not the only direct type that could benefit from `@StateObject`. For example, `ForEachStore`, `SwitchStore` and `IfLetStore`'s implementations currently rely on `WithViewStore`, but one can imagine bespoke optimizations based on `@StateObject`. For example `ForEachStore` could return stable `Store` instances for a given `ID`, bringing it closer to `ForEach` behavior over a collection of value types.
We can of course continue to base these types on `ObservedObject`, and try to shoehorn these improvements as we can but it looks like a sub-optimal solution. Another approach would be to cut a legacy UI library that could be imported by users who need to support older OS's, exposing `Legacy.WithViewStore`, `Legacy.ForEachStore`, but also more subtle helpers like `.alert` and `.confirmationDialog` that are also creating `AnyView`s. As SwiftUI evolves, this problem will present more and more.

**Reply by larryonoff** [reply]
@mbrandonw sorry, I didn't mean inside view builder. I meant as a private variable of WithViewStore. But it looks that it isn't possible.

**Reply by larryonoff** [reply]
According to @tgrapperon it would be better for TCA to drop iOS 13 as planned initially.

**Reply by tgrapperon** [reply]
[Here](https://github.com/pointfreeco/swift-composable-architecture/compare/main...tgrapperon:swift-composable-architecture:extended-support?expand=1) is a very coarse proof-of-concept of a `ComposableArchitectureExtendedSupport` library shipping alongside TCA. I made `WithViewStore` & co iOS 14+, and `.alert` and `.confirmationDialog` iOS 15+. Users who need extended backward compatibility should import `ComposableArchitectureExtendedSupport` and use `Legacy` to namespace their calls. For example, the `AlertAndConfirmationDialog` study would become:
```swift
struct AlertAndConfirmationDialogView: View {
  let store: StoreOf<AlertAndConfirmationDialog>

  var body: some View {
    Legacy.WithViewStore(self.store, observe: { $0 }) { viewStore in
      Form {
        Text("Count: \(viewStore.count)")
        Button("Alert") { viewStore.send(.alertButtonTapped) }
        Button("Confirmation Dialog") { viewStore.send(.confirmationDialogButtonTapped) }
      }
    }
    .legacy.alert(
      self.store.scope(state: \.alert),
      dismiss: .alertDismissed
    )
    .legacy.confirmationDialog(
      self.store.scope(state: \.confirmationDialog),
      dismiss: .confirmationDialogDismissed
    )
  }
}
```
The idea would be to try to keep a concept in core TCA as long as it doesn't force to make concessions, and to translocate it to the `ExtentedSupport` sister library otherwise. This should allow a more fine-grained evolution in the future.

**Reply by sberrevoets** [reply]
I don't have much skin in the game here (nor am I fully up to speed on all the latest features), but maybe it makes sense to provide all the goodies of 1.0 to everyone, and then to make immediately make 1.1 (or, less ideally, 2.0) available that increases the deployment target? That way people that can't drop iOS 13 still can benefit from everything 1.0 brings, but doesn't also force them to drop iOS 13 support.

I'm admittedly not entirely sure what the difference is between the last 0.x.y version and 1.0 though—if it's just making the final renames, defining hard deprecations, and shedding some legacy stuff my suggestion makes less sense (but then I figure folks should also be OK with being on the last pre-1.0 release?).

**Reply by tgrapperon** [reply]
@sberrevoets, this is related to [semantic versioning](https://semver.org). If you follow this scheme, major version numbers like `1.0.0` are boundaries to breaking changes. 

SwiftPM allows to exploit this convention: when you specify `.upToNextMajor(from: 2.3.4)` as some dependency's version, it will automatically accept versions up to `3.0.0` excluded (that is `2.x.y`). So, in the case of TCA, a team that will not be able to make the breaking changes will be able to select `.upToNextMajor(from: 0.33.0)` for example. It will use at least TCA `v0.33.0`, and allow update up to the last `0.something` version (`v0.56.4` for example), but it will not allow itself to go beyond and select the version `1.0.0`[^1].

[^1]: There is also an `.upToNextMinor(from:)` to do the same based on the "middle" number.

The version `1.0` also bears an important symbolic value, as it sign the end of a "beta" phase of a library[^2], and that the API is considered stable enough[^3]. 

[^2]: It means that it is "ReAdy FoR ProDucTiOn", even if it's using SwiftUI which notoriously isn't 😉.
[^3]: TCA is already behaving like it had a `1.0`, in the sense that a particular care was taken to keep changes backward compatible, but this is not a requirement of semantic versioning for versions `0.x.y` which are considered not stable.

**Reply by mbrandonw** [reply]
> I'm admittedly not entirely sure what the difference is between the last 0.x.y version and 1.0 though—if it's just making the final renames, defining hard deprecations, and shedding some legacy stuff my suggestion makes less sense (but then I figure folks should also be OK with being on the last pre-1.0 release?).

Yes, the 1.0 will only consist of renames, deletion of some deprecated code, and (potentially) dropping iOS 13 / Swift 5.6. The last 0.x release will contain all of the tools and features described in the original post, _and_ will still support iOS 13 / Swift 5.6.

**Reply by mbrandonw** [reply]
> [Here](https://github.com/pointfreeco/swift-composable-architecture/compare/main...tgrapperon:swift-composable-architecture:extended-support?expand=1) is a very coarse proof-of-concept of a `ComposableArchitectureExtendedSupport` library shipping alongside TCA.

Hey @tgrapperon, that's a fun exploration! Looks like it could be a really great tool for making breaking changes in a much softer way.

**Reply by tgrapperon** [reply]
@mbrandonw Thanks! A few comments after thinking a bit more about it. It indeed probably needs to ship with TCA itself as I proposed, and not as a separate library to avoid having to sync releases of two libraries.
Furthermore, this kind of namespace should allow a nice "fall-forward" when users finally drop older OS's, as they will only have to remove the `legacy` (or whatever it's called) namespace to rejoin mainstream variants if the APIs are still close enough (they should naturally be, or one could keep them in sync explicitly). This should be able to be signalled automatically and up to the method level using tight deprecations. For example, if `Legacy.WithViewStore` is deprecated starting from iOS 14, it will automatically present a message inviting to switch back to `WithViewStore` as soon as the user will drop support for iOS 13. It can even be an automatic fix using `renamed:`!

**Reply by Alex293** [reply]
I will have a look soon at your experiment, [this](https://davedelong.com/blog/2021/10/09/simplifying-backwards-compatibility-in-swift/) came to my mind when reading your comment. Basically we could have `WithViewStore.backport` nice looking back port code where you just drop the .backport when ready.

**Reply by Alex293** [reply]
The idea originally came from RxSwift code base IIRC with `view.rx` extensions

**Reply by tgrapperon** [reply]
The problem with this is that you would need to make `WithViewStore` visible to iOS 13, so you can't really use suffixes directly in this case. But indeed, it is the same idea.

**Reply by JimRoepcke** [reply]
@mbrandonw you said:

> So, we'd like to drop support for iOS 13 and Swift 5.6 in the 1.0 release of TCA. Two things to keep in mind if you need to support iOS 13 or Swift 5.6:

Would the minimum deployment target be iOS 14 and Swift 5.7, or are you thinking of being even more aggressive there?

**Reply by stephencelis** [reply]
We're thinking we should be able to maintain iOS 13 compatibility, but would like to require Swift 5.7 if possible.

**Comment by davdroman** [comment]
Is there hope for native CasePaths to make it into the library in 1.0?

**Reply by stephencelis** [reply]
Do you mean into the language? We'd love for Swift to ship with case paths, but we don't think it will before our library is 1.0, unfortunately!

**Reply by davdroman** [reply]
@stephencelis that's a bummer 😞

As I understand https://github.com/apple/swift/pull/58940 is the latest in terms of implementation, but we're missing a pitch to kickstart the discussion within the community. Is this initial pitch something you may have enough bandwidth to put out before the end of the year, or is it off the cards completely?

Update: "ask and you shall receive" I guess 😄 https://forums.swift.org/t/enum-case-keypaths/60899

**Comment by stephencelis** [comment]
For those watching, we merged #1478 yesterday, which will hard-deprecate any instances of `Reducer` in the next release, and will give folks a chance to rename their legacy reducers to `AnyReducer` more easily.

@tgrapperon has also done some corresponding work for the `Effect` type in #1471, which we will merge soon if there are no concerns. This means next update you will be able to get a head start on things by converting instances of `Effect<Action, Never>` to `EffectTask<Action>` in your `reduce` signatures. Then, when we introduce the 1.0-prerelease branch, you'll be able to rename `EffectTask` back to `Effect` for an interface we consider 1.0-stable. Sorry for the churn, but we hope you'll be happy with the final result :smile:

**Reply by tylerjames** [reply]
Sorry but what's the difference between a "hard-deprecate" and just a normal or "soft (?) deprecate"?

**Reply by davdroman** [reply]
@tylerjames AFAIK:

- soft-deprecate: doesn't emit warnings, only warns on autocompletion
- hard-deprecate: emits visible warnings on Xcode

**Reply by mbrandonw** [reply]
Soft deprecated also shows up in docs.

It's just a small step we can take to guide people to the correct, future API before we are ready to fully deprecate.

**Comment by stephencelis** [comment]
We just posted a side discussion about another change we'd like to make for 1.0: [Deprecating `Effect.fireAndForget` and `Effect.task` in favor of `Effect.run`](https://github.com/pointfreeco/swift-composable-architecture/discussions/1520)

Let us know what you think!

**Comment by stephencelis** [comment]
Another update! We just released [0.43.0](https://github.com/pointfreeco/swift-composable-architecture/releases/tag/0.43.0) as the first step mentioned in this post.

**Comment by tgrapperon** [comment]
Hot take: It seems we have a unique and time-limited opportunity to be a little bolder with new APIs and experiment with a few things that were kept apart to avoid increasing the surface too much. I was for example thinking about generalized `ForEachReducer` and `ForEachStore` over generic collections, `VoidTaskResult`, etc.
The idea would be to introduce them as soft-deprecated APIs and label them as experimental, so informed users could give them a spin until 1.0 lands. Whether or not they would make the cut could be determined near the end of the beta period.

Furthermore, an hypothetical `ComposableArchitectureExtendedSupport` module indirectly invites to a `ComposableArchitectureExperimental` one where features could be incubated in the same spirit as Swift does with `swift-collections` or `swift-algorithms`, but more fluently without the hassle of having multiple repositories to keep in sync. APIs from this module would be soft-deprecated and without guaranteed backward compatibility, and could be promoted to `ComposableArchitecture` at some point, or be deprecated without backward support if required. Of course, having a `ComposableArchitectureExperimental` module wouldn't mean that any idea would be eligble, but it could allow maturing complex API's in the field without creating friction for more conservative users who could limit themselves to `ComposableArchitecture` and `ComposableArchitectureExtendedSupport`.

**Reply by Alex293** [reply]
Yes please ! There was a lot of nice ideas during the last two years that stayed in forks or even branches here but it's hard to keep up. Having a central place for experiments would greatly simplify tracking those and I think could boost contributions / early reviews. This also would provide more context to ppl with new idea of what was already tried and or rejected.

**Reply by mbrandonw** [reply]
Hi @tgrapperon, I think we can definitely start shipping more experimental APIs, and Stephen and I have discussed a few we'd like to try out. I'm imagining we would just ship the experiments directly in the library hidden behind `@_spi` and we could run a "beta" period for the experiment like we have done with concurrency and `ReducerProtocol`.

But I'm not sure that is related to the 1.0 release. In fact, it might be better to kick off the first experiments _after_ 1.0 to minimize churn.

**Reply by tgrapperon** [reply]
Thanks for the feedback @mbrandonw. I think there were two subjects in my comment, so I'll elaborate more distinctly

### Pre 1.0
By semver, the version 1.0 is potentially a breaking change version. We've already planned a few with the `Effect` and `Reducer` renamings, and maybe making `WithViewStore` iOS 14+.
But also by semver, what's in v1.0 is assumed to be supported without requiring source changes up to 2.0 at least. In the same spirit, what's in v0.99 is not assumed to be supported from 1.0 and after. So we have an opportunity to cut lose a few API's from 0.99 to 1.0 without breaking any implicit contract, and my idea was to exploit this and experiment "without consequences" in the upcoming months, with new API's that aren't assumed to make the 1.0 cut by default. Because of the way TCA is organized today, the simplest would have probably been to introduce the API as soft deprecations (in order to annotate them with a message). They can also be shadowed by `@_spi`, but I don't think it would have been necessary if soft-deprecated.

### 1.0 and after
I'm assuming that the `ComposableArchitectureExtendedSupport` route is chosen, and that APIs which are unmaintainable without hard compromises are relocated there. According to my quick proof of concept with this module, `@_spi` was extremely useful to allow "non-public" APIs' to cross the module's boundary. I feel that it should remain this way, and that `@_spi` annotations should be reserved for internal uses. There's already a lot to digest when coming to TCA, and I'm not sure that `@_spi` is user-friendly enough. It is odd, scary for some folks, and it doesn't perform any validation (`@_spi(Inertal) ComposableArchitecture` compiles without issues for example). I also don't know what are the side effects down the line for consumers of `library-evolution` builds. That's why I felt that `ComposableArchitectureExperimental` would make the intent clearer and more user-friendly. Experimental APIs' would be gathered into their own doc section instead of being sprinkled across the stable ones, and it would be easier to track their evolution toward stable TCA if any. I'm also slightly worried about the increased complexity of `@_spi` during beta periods, as the `ComposableArchitectureExperimental` and its boundary can also be used as a staging area for new stuff.

Of course, one can do without `ComposableArchitectureExtendedSupport` and `ComposableArchitectureExperimental` and use namespaces or `@_spi` in `ComposableArchitecture`, but I feel that the intent is less clear.

**Comment by theMuslimDev** [comment]
How long will it roughly take to get the beta up for navigation? There some old UIKit views I currently have that I'm looking to refactor to SwiftUI but lack of navigation support is making me second guess using TCA for now (well just for that feature, still using TCA in other parts of the app)

**Reply by stephencelis** [reply]
@Muhammed9991 Still some weeks away, but we're working actively on it!

**Reply by strangelearning** [reply]
I might be the only one that would enjoy this but it could be cool to release some kind of behind the scenes of you and @mbrandonw  jamming on coming up with the solutions you eventually do. 

I’d happily pay an extra $10 a month in some kind of Patreon to see what goes into making the sausage, so to speak. 

The solutions you both arrive at are pristine and amazing in their clarity, it would be awesome to learn some of what goes into making that happen!

**Reply by davdroman** [reply]
@strangelearning yes, I've had this thought too. An informal coding/brainstorming live stream where we just get to watch how you zero in on each solution would be an amazing add-on to Point-Free series IMO.

**Reply by Sajjon** [reply]
I would totally pay to watch a coding session! And eat popcorn!

**Reply by strangelearning** [reply]
I think replying tags you but just in case it didn’t, cc @stephencelis

**Reply by ibrahimkteish** [reply]
+1

**Reply by theMuslimDev** [reply]
With all these new navigations videos on point free. Looks like the new navigation stuff will be iOS 16 onwards 🥲 @stephencelis @mbrandonw

**Reply by atimca** [reply]
Maybe it'll be possible to integrate something like this [SwiftUI backport](https://github.com/shaps80/SwiftUIBackports/blob/main/Sources/SwiftUIBackports/Shared/Navigation/NavigationDestination.swift)? User wise it won't affect too many people, majority uses iOS 16 anyhow.

**Reply by mbrandonw** [reply]
@Muhammed9991 not at all! We will have tools for iOS <16 and hopefully even UIKit.

The newest series of navigation episodes is just to complete the series we started 1 year ago since new navigation APIs were introduced. We want to have a collection of episodes that completely covers all SwiftUI navigation tools.

Also, thanks everyone for sharing your thoughts about more "behind the scenes" material from us. We hope to experiment with more of that in 2023! 😁

**Reply by theMuslimDev** [reply]
Phew 😮‍💨. I thought it was psychological trick to mentally prepare everyone 😂. Patiently waiting for it. I've stopped refactor work for a feature from UIKit to SwiftUI in anticipation of this. That particular feature is all about navigation and I really really want to use TCA with it.

Btw no pressure

**Reply by theMuslimDev** [reply]
@Atimca Yeah that's the dream everyone updates straightaway. 

I think I read somewhere iOS 15 was more popular currently than iOS 16. Because Apple found some vulnerabilities in iOS 15. Resulting lots of companies "recommending" there employees to upgrade to latest iOS 15 version (before iOS 16 came out of course). So even users reluctant to upgrade there OS had to and now they're not.

**Reply by tylerjames** [reply]
Good to hear that UIKit and iOS <16 are still in scope. Unless I start some side projects of my own I probably won't be able to use iOS 16 until 2025. I do wish Apple release some official backports

**Reply by theMuslimDev** [reply]
Yep. Not sure what the reasoning for only supporting certain features in latest iOS. They even back ported swift concurrency from iOS 15 to iOS 13. Maybe more stuff can be back ported as well. Well maybe they can but hopefully they will.

**Reply by tylerjames** [reply]
I think that comes down to a difference between Swift, the language, and Apple frameworks that use Swift.

**Comment by j-j-m** [comment]
Will the new navigation features be applicable to UIKit apps?

**Reply by stephencelis** [reply]
The navigation features at the reducer-level should be universally usable in SwiftUI, UIKit, and more. We only plan on shipping view helpers for SwiftUI, but the existing `store.ifLet` method should continue to work well for UIKit navigation.

**Comment by clayellis** [comment]
Has there been any consideration given to this shape of reducer as part of 1.0?

```swift
func reduce(state: inout State, action: Action) async -> Action?
```

Why not just return optional actions asynchronously rather than returning a task? Then instead of returning `.none`, just `return nil`. And instead of:

```swift
case .someAction:
    return .task { 
        let result = await dependency.asyncWork()
        return .otherAction(result)
    }
```

```swift
case .someAction:
    let result = await dependency.asyncWork()
    return .otherAction(result)
```

I suspect it has something to do with cancellation?

**Reply by davdroman** [reply]
AFAIK it's because of [Custom Executors](https://forums.swift.org/t/support-custom-executors-in-swift-concurrency/44425) not being quite there yet in terms of implementation. Cancellation should be possible to do via proper custom executor support.

I believe the other half of the picture was to make `Store`s actors as well which would be super neat for safe `state` mutation.

**Reply by clayellis** [reply]
So this is more like a 2.0 change then?

**Reply by stephencelis** [reply]
I don't think this signature is possible for a couple reasons, but if you explore and find something promising, please do share!

First, I think exposing `inout State` to an async function will introduce problems when it comes to long-living effects, like timers and delays, and the ability to continue to send actions to the store while these effects are in-flight.

Second, returning `Action?` means an action can only ever kick off zero or one more actions, which makes defining things like timers more difficult.

We're definitely open to interesting explorations of TCA "from scratch" with all of Swift's modern tools, though.

**Comment by theMuslimDev** [comment]
Once 1.0 version is released. Are you guys planning on doing a video on how to use composable architecture again? I found the original video very helpful. I believe it was the 100th. Very in-depth. I think people will find this extremely useful @stephencelis @mbrandonw

**Reply by stephencelis** [reply]
Yup! That's definitely something we'd like to cover from scratch now that the library has evolved pretty far from its origins.

**Comment by andrewjroberts** [comment]
anyone else spamming refresh seeing if there's going to be a new pinned discussion this weekend?  🤞🤠!!

**Reply by acosmicflamingo** [reply]
To be honest, I've been looking through all of PointFree's GitHub repos and seeing what branches have been recently updated (cough cough navigation cough cough). Unfortunately for us, Brandon and Stephen are covering their tracks pretty well and I have no idea what is coming our way :(

**Reply by davdroman** [reply]
I think next week will most likely be off whilst they refine navi-cough-cough. I'm personally not holding my breath until next weekend. We could be pleasantly surprised though 🤷‍♂️

**Reply by acosmicflamingo** [reply]
Oh that's a good point. I'll just have to catch up on some of their other videos until then 😃

**Reply by andrewjroberts** [reply]
lets gooooooo team TCA (rooting for you guys, however long it takes)

**Comment by pitt500** [comment]
I took my time to read all the threads and the ideas from the community and it's really amazing to see all the attention to TCA evolution. Great job to all the people making this possible! 😃

@stephencelis @mbrandonw I have a question, Do you have any plans to make TCA library platform agnostic for 1.0 or X.0 ?
I read that some time ago and based on the comments I don't think that will come in 1.0, but just curious. Thanks!

**Reply by mbrandonw** [reply]
Hi @pitt500, we aren't going to get platform independence for the library in 1.0, but it could possibly be a goal for 2.0. With the new macro features, `Observation` protocol and concurrency tools, it is getting easier and easier to depend on fully native Swift constructs.

**Reply by pitt500** [reply]
That sounds cool. I'm not familiar with Swift Macros yet, Do you have a link that I can take a look? Thanks in advance.

EDIT: I found this btw https://github.com/DougGregor/swift-macro-examples

**Reply by mbrandonw** [reply]
And here's a dashboard of current state of the various proposals: https://gist.github.com/DougGregor/de840fcf6d6f307792121eee11c0da85
